llama.cpp

```diff
@@ -16427,6 +16428,18 @@ struct llama_context * llama_new_context_with_model(
             }
             ctx->backends.push_back(backend);
         }
+#endif
+#if defined(GGML_USE_RPC)
+        for (int i = 0; i < (int)model->rpc_servers.size(); i++) {
+            const char * endpoint = model->rpc_servers[i].c_str();
+            ggml_backend_t backend = ggml_backend_rpc_init(endpoint);
+            if (backend == nullptr) {
+                LLAMA_LOG_ERROR("%s: failed to initialize RPC to '%s'\n", __func__, endpoint);
+                llama_free(ctx);
+                return nullptr;
+            }
+            ctx->backends.push_back(backend);
+        }
```

Comment:
```suggestion
        if (model->n_gpu_layers > 0) {
            for (const auto & endpoint : model->rpc_servers) {
                ggml_backend_t backend = ggml_backend_rpc_init(endpoint);
                if (backend == nullptr) {
                    LLAMA_LOG_ERROR("%s: failed to initialize RPC to '%s'\n", __func__, endpoint);
                    llama_free(ctx);
                    return nullptr;
                }
                ctx->backends.push_back(backend);
            }
        }
```

#7711 made me realize that `ggml_backend_rpc_init` cannot fail because it never attempts a connection. Then it returns a `NULL` buffer type in `get_default_buffer_type`, which is never supposed to happen.

Reply:
should we also check for `LLAMA_SPLIT_MODE_LAYER` here in addition to `n_gpu_layers > 0`?

Reply:
All modes would require initializing the RPC backend. It could be improved to check that the backend is actually used in the model before instantiating it, but mostly the goal here is to avoid issues with `-ngl 0`. I think that the bug on master that causes a RPC backend with a empty name to be initialized always is not present here.